File: packages/sdk/go/fingerprint_test.go
package bugwatch

import (
	"testing"
)

func TestGenerateFingerprintConsistency(t *testing.T) {
	event := &ErrorEvent{
		Exception: &ExceptionInfo{
			Type:  "RuntimeError",
			Value: "something broke",
			Stacktrace: []StackFrame{
				{Filename: "main.go", Function: "main.handleRequest", InApp: true},
			},
		},
	}

	fp1 := GenerateFingerprint(event)
	fp2 := GenerateFingerprint(event)

	if fp1 != fp2 {
		t.Errorf("Fingerprint should be consistent: got %s and %s", fp1, fp2)
	}

	if fp1 == "" {
		t.Error("Fingerprint should not be empty")
	}
}

func TestGenerateFingerprintDifferentInputs(t *testing.T) {
	event1 := &ErrorEvent{
		Exception: &ExceptionInfo{
			Type:  "RuntimeError",
			Value: "error one",
			Stacktrace: []StackFrame{
				{Filename: "main.go", Function: "main.handler", InApp: true},
			},
		},
	}

	event2 := &ErrorEvent{
		Exception: &ExceptionInfo{
			Type:  "TypeError",
			Value: "error two",
			Stacktrace: []StackFrame{
				{Filename: "api.go", Function: "api.process", InApp: true},
			},
		},
	}

	fp1 := GenerateFingerprint(event1)
	fp2 := GenerateFingerprint(event2)

	if fp1 == fp2 {
		t.Errorf("Different events should produce different fingerprints: both got %s", fp1)
	}
}

func TestGenerateFingerprintNoException(t *testing.T) {
	event := &ErrorEvent{}

	fp := GenerateFingerprint(event)
	if fp == "" {
		t.Error("Fingerprint should not be empty even without exceptions")
	}
}

func TestGenerateFingerprintUsesFirstInAppFrame(t *testing.T) {
	event1 := &ErrorEvent{
		Exception: &ExceptionInfo{
			Type:  "Error",
			Value: "oops",
			Stacktrace: []StackFrame{
				{Filename: "vendor/lib.go", Function: "lib.Do", InApp: false},
				{Filename: "main.go", Function: "main.run", InApp: true},
				{Filename: "handler.go", Function: "handler.serve", InApp: true},
			},
		},
	}

	event2 := &ErrorEvent{
		Exception: &ExceptionInfo{
			Type:  "Error",
			Value: "oops",
			Stacktrace: []StackFrame{
				{Filename: "vendor/lib.go", Function: "lib.Do", InApp: false},
				{Filename: "other.go", Function: "other.run", InApp: true},
			},
		},
	}

	fp1 := GenerateFingerprint(event1)
	fp2 := GenerateFingerprint(event2)

	if fp1 == fp2 {
		t.Error("Events with different first in-app frames should have different fingerprints")
	}
}

File: packages/sdk/java/core/src/test/java/dev/bugwatch/FingerprintTest.java
package dev.bugwatch;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class FingerprintTest {

    @Test
    void testGenerateProducesConsistentHash() {
        ErrorEvent event = createTestEvent("java.lang.RuntimeException", "test error");
        String fp1 = Fingerprint.generate(event);
        String fp2 = Fingerprint.generate(event);

        assertNotNull(fp1);
        assertFalse(fp1.isBlank());
        assertEquals(fp1, fp2, "Fingerprint should be deterministic");
    }

    @Test
    void testDifferentExceptionsProduceDifferentFingerprints() {
        ErrorEvent event1 = createTestEvent("java.lang.RuntimeException", "error one");
        ErrorEvent event2 = createTestEvent("java.lang.IllegalStateException", "error two");

        String fp1 = Fingerprint.generate(event1);
        String fp2 = Fingerprint.generate(event2);

        assertNotEquals(fp1, fp2, "Different exceptions should produce different fingerprints");
    }

    @Test
    void testSameExceptionTypeSameMessageSameFingerprint() {
        ErrorEvent event1 = createTestEvent("java.lang.NullPointerException", "null ref");
        ErrorEvent event2 = createTestEvent("java.lang.NullPointerException", "null ref");

        String fp1 = Fingerprint.generate(event1);
        String fp2 = Fingerprint.generate(event2);

        assertEquals(fp1, fp2, "Same exception type and message should produce the same fingerprint");
    }

    @Test
    void testEmptyEventProducesFingerprint() {
        ErrorEvent event = new ErrorEvent();
        String fp = Fingerprint.generate(event);
        assertNotNull(fp);
        assertFalse(fp.isBlank());
    }

    @Test
    void testFingerprintIsSha256Length() {
        ErrorEvent event = createTestEvent("Exception", "test");
        String fp = Fingerprint.generate(event);
        // SHA-256 produces 64 hex characters
        assertEquals(64, fp.length());
    }

    private ErrorEvent createTestEvent(String type, String message) {
        ErrorEvent event = new ErrorEvent();
        List<StackFrame> frames = List.of(
                new StackFrame("Test.java", "com.example.Test.run", 42, null, "Test.java", true, "com.example")
        );
        event.setException(new ExceptionInfo(
                type,
                message,
                frames
        ));
        return event;
    }
}
